design/external-authorization-design.md

```diff
@@ -0,0 +1,670 @@
+     |                            | | redirection   |           |                   |              |
+     |     Authorization redirect | |---------------|           |                   |              |
+     |<---------------------------|                             |                   |              |
+     |                            |                             |                   |              |
+     | Sign in                    |                             |                   |              |
+     |----------------------------------------------------------------------------->|              |
+     |                            |                             |                   |              |
+     |                            |                          Authorization response |              |
+     |<-----------------------------------------------------------------------------|              |
+     |                            |                             |                   |              |
+     | Resend HTTP request        |                             |                   |              |
+     |--------------------------->|                             |                   |              |
+     |                            |                             |                   |              |
+```
+
+## Alternatives Considered
+
+1. Contour could install itself as the authorization server.
+   This could remove some of the limitations of the Envoy configuration structure at the cost of more complexity in Contour.
+1. Integrate external authorization directly into `HTTPProxy`.
+   This increases the complexity of the `HTTPProxy` structure and makes it difficult to re-use the same authorization service acrtoss multiple proxies.
+   A separate CRD gives better opportunities to expose useful operational status.
+   Integrating specific authorization parameters into `HTTPProxy` prevents operators implementing their own authorization flows.
+
+## Security Considerations
+
+- Most HTTP authorization protocols depend on HTTPS to keep headers and URL query parameters private.
+  Supporting only authorization on HTTPS, Contour reduces the changes of misconfiguring this.
+- HTTPS sessions between Envoy and the authorization server are required.
+  TLS validation can be configured if necessary (should be recommended),
+- Authorization services can run in separate Kubernetes namespaces with limited privilege.
+
+## Compatibility
+
+Contour can currently support v2 of the Envoy GRPC authorization protocol.
+When Contour moves to v3 of the Envoy API, it can support both v2 and v3 authorization protocols.
+
+The `SupportService` CRD supports GRPC authorization support services, but not HTTP ones.
+
+## Implementation
+
+The Contour implementation provides a framework, but isn't useful until authorization services are available for the protocols that Contour operators need.
```

Comment:
Are we (Contour maintainers) realistically going to be building & maintaining all of the auth services? Or are there other obvious parties who would be interested in collaborating on these or building new ones?

I'm just wondering, if practically we're going to be building and maintaining all of them, if that might change how we'd approach the framework design.

Reply:
I don't know the answers to that. However, it is a requirement that third parties be able to add their own authorization.

Reply:
We should build whatever is needed to make OIDC+OAuth2 work really well.